pkgs/ffigen/lib/src/code_generator/typealias.dart

```diff
@@ -19,6 +19,7 @@ class Typealias extends BindingType {
   final Type type;
   String? _ffiDartAliasName;
   String? _dartAliasName;
+  String? _wrapperAliasName;
```

Comment:
> If we already have a class that extends `ObjCBlockBase` with the right type arguments, why return a `ObjCBlockBase` in the public API at all? Can we just only use the class in the public API? Why not?

I'll give a quick answer to this question, and leave the rest of the comments for tomorrow. This is pretty much the whole point of the change. The bug I'm trying to fix is when an ObjC child class overrides a method on its parent class and accepts/returns different block types to the parent (it obeys the subtyping rules like in the block_inherit test, but the block type is not identical).

If our codegenned wrapper API accepts/returns the wrapper block types, it doesn't matter whether we encode the function signature in `ObjCBlockBase`, it will still violate Dart's subtyping rules. The only way it works is if we accept/return `ObjCBlockBase` directly.

Reply:
Ah right!

Yeah then I think I agree with you that the cleanest solution would be to never have instantiations of the wrapper type:

> On that last point, a more drastic breaking change we could do is to change the wrapper type to be an unconstructible class that just holds static methods for creating `ObjCBlockBase<...>`. Users would never hold references to the wrapper type, just `ObjCBlockBase<...>`. We'd have to change the call method on the wrapper type to a static method that takes the block as the first arg: `static Ret invoke(ObjCBlockBase<...> block, args...)`. Slightly cleaner since now there's just `ObjCBlockBase<...>`, but it's a bigger breaking change, and makes calling the block from Dart less ergonomic.

Those wrapper classes would just hold static methods, and they don't need to extend anything. That would definitely create less confusion for users than having a class subtype `ObjCBlockBase`. Because with the subtyping I believe users are going to have questions which of the two types they should be using.

Reply:
Now I'm wondering if it would be possible to add a code genned call operator to ObjCBlockBase, on a per-signature basis, using extension methods.

Reply:
Can't use extension methods because the C types are not always enough to distinguish the different kinds of block. A block that takes a nullable Object pointer and a block that takes a non-nullable pointer will have the same C types (but different Dart types).

Reply:
Gonna go with option 2 since it's a lot more readable.

Reply:
Done.